pkg/cryptoauthlib/psa_se_driver/psa_atca_se_driver.c

```diff
@@ -0,0 +1,561 @@
+{
+    ctx->device = dev;
+    ctx->key_id = key_slot;
+    ctx->key_block = 0;
+}
+
+psa_status_t atca_cipher_setup(psa_drv_se_context_t *drv_context,
+                               void *op_context,
+                               psa_key_slot_number_t key_slot,
+                               psa_algorithm_t algorithm,
+                               psa_encrypt_or_decrypt_t direction)
+{
+    ATCADevice dev = (ATCADevice)drv_context->transient_data;
+    psa_se_cipher_context_t *ctx = (psa_se_cipher_context_t *)op_context;
+
+    /* Only device type ATECC608 supports AES operations */
+    if (dev->mIface.mIfaceCFG->devtype != ATECC608) {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    /* This implementation is for demonstration and currently only supports AES ECB encryption */
+    if (!ALG_IS_SUPPORTED(algorithm)) {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    switch (algorithm) {
+    case PSA_ALG_CBC_NO_PADDING:
+        atca_cbc_setup(&ctx->drv_ctx.atca_aes_cbc, dev, key_slot);
+        break;
+    default:
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+    ctx->direction = direction;
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atca_cipher_set_iv(void *op_context,
+                                const uint8_t *p_iv,
+                                size_t iv_length)
+{
+
+    atca_aes_cbc_ctx_t *ctx = &((psa_se_cipher_context_t *)op_context)->drv_ctx.atca_aes_cbc;
+
+    if (iv_length != ATCA_MAX_IV_LEN) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    memcpy(ctx->ciphertext, p_iv, iv_length);
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atca_cipher_update(void *op_context,
+                                const uint8_t *p_input,
+                                size_t input_size,
+                                uint8_t *p_output,
+                                size_t output_size,
+                                size_t *p_output_length)
+{
+    psa_se_cipher_context_t *ctx = (psa_se_cipher_context_t *)op_context;
+    ATCA_STATUS status = ATCA_EXECUTION_ERROR;
+    size_t input_offset = 0;
+    size_t output_offset = 0;
+
+    for (size_t data_block = 0; data_block < (input_size / AES_128_BLOCK_SIZE); data_block++) {
+        output_offset = data_block * AES_128_BLOCK_SIZE;
+        input_offset = data_block * AES_128_BLOCK_SIZE;
+        if (ctx->direction == PSA_CRYPTO_DRIVER_ENCRYPT) {
+            status = atcab_aes_cbc_encrypt_block(&ctx->drv_ctx.atca_aes_cbc, p_input + input_offset,
+                                                 p_output + output_offset);
+        }
+        else {
+            status = atcab_aes_cbc_decrypt_block(&ctx->drv_ctx.atca_aes_cbc, p_input + input_offset,
+                                                 p_output + output_offset);
+        }
+    }
+
+    if (status != ATCA_SUCCESS) {
+        DEBUG("ATCA Error: %d\n", status);
+        return atca_to_psa_error(status);
+    }
+
+    *p_output_length += input_size;
```

Comment:
This leaves the `p_output_length` in an unclear length, when the function is aborted earlier. E.g. three blocks might have been decrypted already when an error is encountered.

Reply:
Done